File: backend/proxy.py
import requests
import json

INPUT_PROXY_FILE = "backend/src/mazproxy.txt"
OUTPUT_PROXY_FILE = "backend/src/mazcekproxy.txt"

API_URL = "https://check.mazlana.biz.id/api/v1"

def read_proxy_list():
    proxies = []
    with open(INPUT_PROXY_FILE) as f:
        for line in f:
            if not line.strip():
                continue
            parts = line.strip().split(",")
            if len(parts) >= 2:
                ip, port = parts[0], int(parts[1])
                country = parts[2] if len(parts) > 2 else "Unknown"
                org = parts[3] if len(parts) > 3 else "Unknown"
                proxies.append((ip, port, country, org))
    return proxies

def check_proxy(ip, port):
    try:
        resp = requests.get(API_URL, params={"ip": ip, "port": port}, timeout=10)
        if resp.status_code != 200:
            return {"error": f"HTTP {resp.status_code}"}
        return resp.json()
    except Exception as e:
        return {"error": str(e)}

def main():
    proxy_list = read_proxy_list()
    print(f"🔍 Mengecek {len(proxy_list)} proxy...\n")

    all_active = []
    for idx, (ip, port, country, org) in enumerate(proxy_list, 1):
        meta = check_proxy(ip, port)
        if meta and not meta.get("error") and meta.get("delay", 0) > 0:
            asn = meta.get("asn", "Unknown")
            asOrganization = meta.get("asOrganization", "Unknown")
            colo = meta.get("colo", "Unknown")
            latitude = meta.get("latitude", "Unknown")
            longitude = meta.get("longitude", "Unknown")
            delay = meta.get("delay", 0)

            line = f"{ip},{port},{country},{org},{asn},{asOrganization},{colo},{latitude},{longitude},{delay}"
            all_active.append(line)

            print(f"[{idx}/{len(proxy_list)}] ✅ Alive {ip}:{port} {delay}ms {country} {asOrganization}")
        else:
            print(f"[{idx}/{len(proxy_list)}] ❌ Dead {ip}:{port} ({meta.get('error') if meta else 'unknown'})")

    # simpan hasil
    with open(OUTPUT_PROXY_FILE, "w") as out:
        out.write("\n".join(all_active))

    print(f"\nSelesai! Total proxy aktif: {len(all_active)} disimpan ke {OUTPUT_PROXY_FILE}")

if __name__ == "__main__":
    main()

File: backend/proxy.js
// backend/proxy.js
import fs from "fs/promises";

// File input/output
const INPUT_PROXY_FILE = "backend/src/mazproxy.txt";
const OUTPUT_PROXY_FILE = "backend/src/mazcekproxy.txt";

const API_URL = "https://check.mazlana.biz.id/api/v1";
const CONCURRENCY = 50; // jumlah max request bersamaan

// Baca list proxy
async function readProxyList() {
  const proxies = [];
  const text = await fs.readFile(INPUT_PROXY_FILE, "utf-8");
  for (const line of text.split("\n")) {
    if (!line.trim()) continue;
    const parts = line.trim().split(",");
    if (parts.length >= 2) {
      const ip = parts[0];
      const port = parseInt(parts[1]);
      const country = parts[2] || "Unknown";
      const org = parts[3] || "Unknown";
      proxies.push({ ip, port, country, org });
    }
  }
  return proxies;
}

// Call API health check
async function checkProxy(ip, port) {
  try {
    const url = `${API_URL}?ip=${encodeURIComponent(ip)}&port=${port}`;
    const resp = await fetch(url, { timeout: 10000 });
    if (!resp.ok) {
      return { error: `HTTP ${resp.status}` };
    }
    return await resp.json();
  } catch (err) {
    return { error: String(err) };
  }
}

// Batasi concurrency (parallel check)
async function mapLimit(items, limit, fn) {
  const results = [];
  let idx = 0;

  async function worker() {
    while (idx < items.length) {
      const i = idx++;
      results[i] = await fn(items[i], i);
    }
  }

  await Promise.all(new Array(Math.min(limit, items.length)).fill(0).map(worker));
  return results;
}

async function main() {
  const proxyList = await readProxyList();
  console.log(`🔍 Mengecek ${proxyList.length} proxy...\n`);

  const results = await mapLimit(proxyList, CONCURRENCY, async (proxy, i) => {
    const meta = await checkProxy(proxy.ip, proxy.port);

    if (meta && !meta.error && meta.delay && meta.delay > 0) {
      const line = [
        proxy.ip,
        proxy.port,
        proxy.country,
        proxy.org,
        meta.asn || "Unknown",
        meta.asOrganization || "Unknown",
        meta.colo || "Unknown",
        meta.latitude || "Unknown",
        meta.longitude || "Unknown",
        meta.delay || 0,
      ].join(",");

      console.log(
        `[${i + 1}/${proxyList.length}] ✅ Alive ${proxy.ip}:${proxy.port} ${meta.delay}ms ${proxy.country} ${meta.asOrganization}`
      );
      return line;
    } else {
      console.log(
        `[${i + 1}/${proxyList.length}] ❌ Dead ${proxy.ip}:${proxy.port} (${meta?.error || "unknown"})`
      );
      return null;
    }
  });

  const active = results.filter(Boolean);
  await fs.writeFile(OUTPUT_PROXY_FILE, active.join("\n"), "utf-8");

  console.log(`\nSelesai! Total proxy aktif: ${active.length} disimpan ke ${OUTPUT_PROXY_FILE}`);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
